Source/Core/CustomProjection.js

```diff
@@ -0,0 +1,288 @@
+        //>>includeEnd('debug');
+
+        if (!defined(result)) {
+            result = new Cartographic();
+        }
+
+        this._unproject(cartesian, result);
+        return result;
+    };
+
+    function buildCustomProjection(customProjection, url, projectionName) {
+        var fetch;
+        var deferred = when.defer();
+        if (defined(loadedProjectionFunctions[projectionName])) {
+            loadedProjectionFunctions[projectionName](function(project, unproject) {
+                customProjection._project = project;
+                customProjection._unproject = unproject;
+                customProjection._ready = true;
+                deferred.resolve(customProjection);
+            });
+
+            return deferred;
+        }
+
+        if (typeof WorkerGlobalScope !== 'undefined' && self instanceof WorkerGlobalScope) { // eslint-disable-line no-undef
+            importScripts(url); // eslint-disable-line no-undef
+            fetch = when.resolve();
+        } else {
+            fetch = Resource.fetchJsonp({
+                    url : url
+            });
+        }
+
+        fetch = fetch
+            .then(function() {
+                var localCreateProjectionFunctions = createProjectionFunctions; // eslint-disable-line no-undef
+                loadedProjectionFunctions[projectionName] = localCreateProjectionFunctions;
+                localCreateProjectionFunctions(function(project, unproject) {
+                    customProjection._project = project;
+                    customProjection._unproject = unproject;
+                    customProjection._ready = true;
+                    deferred.resolve();
+                });
+
+                return deferred.promise;
+            })
+            .then(function() {
+                return customProjection;
+            });
+
+        return fetch;
+    }
+
+    /**
+     * A function used to generate functions for a custom MapProjection.
+     * This function must be named <code>createProjectionFunctions</code>.
+     * @callback CustomProjection~factory
+     *
+     * @param {Function} callback A callback that takes <code>CustomProjection~project</code> and <code>CustomProjection~unproject</code> functions as arguments.
+     * @example
+     * function simpleProjection(callback) {
+     *     function project(cartographic, result) {
+     *          result.x = cartographic.longitude * 6378137.0;
+     *          result.y = cartographic.latitude * 6378137.0;
+     *          result.z = cartographic.height;
+     *     }
+     *
+     *     function unproject(cartesian, result) {
+     *          result.longitude = cartesian.x / 6378137.0;
+     *          result.latitude = cartesian.y / 6378137.0;
+     *          result.height = cartesian.z;
+     *     }
+     *
+     *     callback(project, unproject);
+     * }
+     */
+
+    /**
+     * A function that projects a cartographic coordinate to x/y/z coordinates in 2.5D space.
```

Comment:
This and the doc below should mention the orientation of the surface and units. Something like "projects the lat/lon to the xy-plane and z is the height above the surface". What are the units? meters?